odl/deform/linearized_deform.py

```diff
@@ -0,0 +1,497 @@
+# Copyright 2014-2016 The ODL development group
+#
+# This file is part of ODL.
+#
+# ODL is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# ODL is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with ODL.  If not, see <http://www.gnu.org/licenses/>.
+"""
+Linearized deformations.
+"""
+
+# Imports for common Python 2/3 codebase
+from __future__ import print_function, division, absolute_import
+from future import standard_library
+from odl.operator.operator import Operator
+import odl
+import numpy as np
+standard_library.install_aliases()
+
+
+__all__ = ('linear_deform', 'deform_grad', 'LinDeforFixedTempOp',
+           'LinDeforFixedTempDeriv', 'LinDeforFixedTempDerivAdj',
+           'LinDeforFixedDispOp', 'LinDeforFixedDispAdj')
+
+
+def linear_deform(template, displacement):
+    """Linearized deformation of a template with a displacement field.
+
+    The template ``I`` is deformed by a displacement field ``v`` as:
+
+        (I, v) --> I(Id + v)
+
+    Parameters
+    ----------
+    template : `DiscreteLpVector`
+        Template to be deformed by a displacement field.
+    displacement : `ProductSpaceVector`
+        The vector field (displacement field) used in the linearized
+        deformation.
+
+    Returns
+    -------
+    deformed_template : `numpy.ndarray`
+        Deformed template as an numpy array.
+
+    Examples
+    --------
+    Deform a 1D function ``template`` that has value ``1`` at 3:rd sample
+    point using a displacement field ``displacement_field`` that has
+    value ``-0.2`` in the 4:th sample point. The outcome should be an
+    array with values ``1`` at 3:rd and 4:th positions.
+
+    >>> import odl
+    >>> space = odl.uniform_discr(0, 1, 5)
+    >>> disp_field_space = odl.ProductSpace(space, space.ndim)
+    >>> template = space.element([0, 0, 1, 0, 0])
+    >>> displacement_field = disp_field_space.element([[0, 0, 0, -0.2, 0]])
+    >>> linear_deform(template, displacement_field)
+    array([ 0.,  0.,  1.,  1.,  0.])
+    """
+    assert template in displacement[0].space
+
+    image_pts = template.space.grid.points()
+    image_pts += np.asarray(displacement).T
+
+    return template.interpolation(image_pts.T, bounds_check=False)
+
+
+def deform_grad(grad_f, displacement):
```

Comment:
The functionality in ``linear_deform`` is shared in both types of deformation operators, fixed displacement and fixed template. It is important that this is consistently done in both operators, so that is why it implemented as a helper function even though it merely consists of two lines.